mlflow/pyfunc/__init__.py

```diff
@@ -656,12 +675,154 @@ def _warn_dependency_requirement_mismatches(model_path):
                 f"Python environment:\n{mismatch_str}"
             )
             _logger.warning(warning_msg)
+            return False
+        else:
+            return True
     except Exception as e:
         _logger.warning(
             f"Encountered an unexpected error ({repr(e)}) while detecting model dependency "
             "mismatches. Set logging level to DEBUG to see the full traceback."
         )
         _logger.debug("", exc_info=True)
+        return e
+
+
+_PYENV_CMD = 'pyenv'
+_VIRTUALENV_CMD = 'virtualenv'
+
+
+_PYENV_HOME = None
+
+
+def _get_or_create_pyenv_home_dir():
+    """
+    Note: pyenv home is a directory SHARED across all REPLs/Users processes.
+    all python installations are put in this directory.
+    On databricks, the path is under root NFS directory, so it can be shared with all cluster nodes.
+    """
+    global _PYENV_HOME
+
+    if _PYENV_HOME is None:
+        nfs_root_dir = get_nfs_cache_root_dir()
+
+        if nfs_root_dir is not None:
+            _PYENV_HOME = os.path.join(nfs_root_dir, 'pyenv_home')
+            os.makedirs(_PYENV_HOME, exist_ok=True)
+        else:
+            _PYENV_HOME = os.path.expanduser('~')
+
+    return _PYENV_HOME
+
+
+_VIRTUAL_ENV_DIR = None
+
+
+def _get_or_create_python_virtual_env_temp_dir():
+    """
+    Note: the python virtual env dir is NOT SHARED across different REPLs/users processes.
+    and when python process exist the directory will be automatically removed.
+    """
+    global _VIRTUAL_ENV_DIR
+    if _VIRTUAL_ENV_DIR is None:
+        nfs_root_dir = get_nfs_cache_root_dir()
+        if nfs_root_dir is not None:
+            # In databricks, the '/local_disk0/.ephemeral_nfs' is mounted as NFS disk
+            # the data stored in the disk is shared with all remote nodes.
+            root_dir = os.path.join(nfs_root_dir, "venvs")
+            os.makedirs(root_dir, exist_ok=True)
+            _VIRTUAL_ENV_DIR = tempfile.mkdtemp(dir=root_dir)
+            # TODO: register deleting _VIRTUAL_ENV_DIR handler when exit
+        else:
+            import atexit
+            import shutil
+            _VIRTUAL_ENV_DIR = tempfile.mkdtemp()
+            atexit.register(shutil.rmtree, _VIRTUAL_ENV_DIR, ignore_errors=True)
+
+    return _VIRTUAL_ENV_DIR
+
+
+_PICKLE_PROTOCOL_FOR_RESTORE_PY_ENV = 3
+
+
+def _setup_restored_python_env(py_version, requirements_file_path):
```

Comment:
Following from https://github.com/mlflow/mlflow/pull/5470/files#r822326583, we should introduce this to `PyFuncBackend` so that it can be used for `mlflow models serve` and Spark UDF.